### ISSUE TITLE
CoreWebView2Environment not working

### ISSUE THREAD

**Issue by cgeier** [MicrosoftEdge/WebView2Feedback#383]
**Issue:** Can't set web cache data folder, using CoreWebView2Environment, if WebView2 control Source property is set. When using the Visual Studio form designer, this property is set. 

If one creates a form named Form1 using the Visual Studio form designer, the following files are created: 

- Form1.cs
- Form1.Designer.cs

Add a WebView2 control to the form named: WebView1

Attempting to set the CoreWeb2Environment, doesn't work. 

```
CoreWebView2Environment webView2Environment = await CoreWebView2Environment.CreateAsync(null, @"C:\Temp\WebView2_Cache", null);

await WebView1.EnsureCoreWebView2Async(webView2Environment);

WebView1.Source = new Uri("about:blank", UriKind.Absolute);
```

If one goes to **Form1.Designer.cs**, and expands `Windows Form Designer generated code
` the following is seen (which is typical):

```
        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            this.WebView1 = new Microsoft.Web.WebView2.WinForms.WebView2();
            this.SuspendLayout();
            // 
            // WebView1
            // 
            this.WebView1.Dock = System.Windows.Forms.DockStyle.Fill;
            this.WebView1.Location = new System.Drawing.Point(0, 0);
            this.WebView1.Name = "WebView1";
            this.WebView1.Size = new System.Drawing.Size(800, 450);
            this.WebView1.Source = new System.Uri("about:blank", System.UriKind.Absolute);
            this.WebView1.TabIndex = 0;
            this.WebView1.Text = "WebView1";
            this.WebView1.ZoomFactor = 1D;
            // 
            // Form1
            // 
            this.AutoScaleDimensions = new System.Drawing.SizeF(6F, 13F);
            this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font;
            this.ClientSize = new System.Drawing.Size(800, 450);
            this.Controls.Add(this.WebView1);
            this.Name = "Form1";
            this.Text = "Form1";
            this.Load += new System.EventHandler(this.Form1_Load);
            this.ResumeLayout(false);

        }
```
If one comments out `this.WebView1.Source = new System.Uri("about:blank", System.UriKind.Absolute);` , setting the CoreWebView2Environment will work.

After the line is commented out, Form1.Designer.cs file looks like the following--which works:

**Form1.Designer.cs**

```
namespace WebView2SetEnvironmentTest
{
    partial class Form1
    {
        /// <summary>
        /// Required designer variable.
        /// </summary>
        private System.ComponentModel.IContainer components = null;

        /// <summary>
        /// Clean up any resources being used.
        /// </summary>
        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
        protected override void Dispose(bool disposing)
        {
            if (disposing && (components != null))
            {
                components.Dispose();
            }
            base.Dispose(disposing);
        }

        #region Windows Form Designer generated code

        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            this.WebView1 = new Microsoft.Web.WebView2.WinForms.WebView2();
            this.SuspendLayout();
            // 
            // WebView1
            // 
            this.WebView1.Dock = System.Windows.Forms.DockStyle.Fill;
            this.WebView1.Location = new System.Drawing.Point(0, 0);
            this.WebView1.Name = "WebView1";
            this.WebView1.Size = new System.Drawing.Size(800, 450);
            //this.WebView1.Source = new System.Uri("about:blank", System.UriKind.Absolute);
            this.WebView1.TabIndex = 0;
            this.WebView1.Text = "WebView1";
            this.WebView1.ZoomFactor = 1D;
            // 
            // Form1
            // 
            this.AutoScaleDimensions = new System.Drawing.SizeF(6F, 13F);
            this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font;
            this.ClientSize = new System.Drawing.Size(800, 450);
            this.Controls.Add(this.WebView1);
            this.Name = "Form1";
            this.Text = "Form1";
            this.Load += new System.EventHandler(this.Form1_Load);
            this.ResumeLayout(false);

        }

        #endregion

        private Microsoft.Web.WebView2.WinForms.WebView2 WebView1;
    }
}
```

**Root cause**:  Once the Source property is set, CoreWebView2 environment can't be set--the Visual Studio form designer sets this property.

It's my understanding that the WebView2 cache, is created in the same folder as the executable, unless another location was specified using CoreWebView2Environment. This can be a problem if the user doesn't have write permissions on the folder that the executable is running from--such as when the application has been installed to %ProgramFiles%. (ex: C:\Program Files\<my application>). There should be a way to set the CoreWebView2Environment when adding the WebView2, control using the Visual Studio form designer, without having to comment out the Source property in Form1.Designer.cs. When the CoreWebView2Environment works correctly, it creates a folder named `EBWebView` within the folder that was specified by the user.

Here's the code for Form1.cs, which works if `this.WebView1.Source = new System.Uri("about:blank", System.UriKind.Absolute);` is commented out in Form1.Designer.cs, as previously described.

**Form1.cs**

```
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using Microsoft.Web.WebView2.Core;
using Microsoft.Web.WebView2.WinForms;
using Microsoft.Web.WebView2;
using System.IO;
using System.Diagnostics;

namespace WebView2SetEnvironmentTest
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            //add event handler for CoreWebView2Ready - before WebView2 control is initialized
            //it's important to not use WebView2 control until CoreWebView2Ready event is thrown
            WebView1.CoreWebView2Ready += WebView1_CoreWebView2Ready;

            //InitializeWebView2Async(WebView1,"http://www.bing.com");
            InitializeWebView2Async(WebView1,"http://www.bing.com", @"C:\Temp\WebView2_Cache");
        
        }

        private async void InitializeWebView2Async(WebView2 wv, string initialSource, string webCacheDir = "")
        {
            //set value
            string tempWebCacheDir = webCacheDir;

            //if user didn't specify a webCacheDir, 
            //use the user's temp folder
            if (String.IsNullOrEmpty(tempWebCacheDir))
            {
                //get fully-qualified path to user's temp folder
                tempWebCacheDir = System.IO.Path.GetTempPath();
            }//if

            CoreWebView2EnvironmentOptions options = null;
            //options = new CoreWebView2EnvironmentOptions("--disk-cache-size=200");
            //options = new CoreWebView2EnvironmentOptions("–incognito ");

            //set WebView2 control cache/data folder. The cache/data folder is used to 
            //store WebView2 cached objects. If not specified, the folder where the executable
            //was started will be used. If the user doesn't have write permissions
            //on that folder, such as C:\Program Files\<your application folder>\,
            //an exception will be thrown 

            CoreWebView2Environment webView2Environment = null;
            //webView2Environment = await CoreWebView2Environment.CreateAsync(@"C:\Program Files (x86)\Microsoft\Edge Dev\Application\85.0.564.8", tempDir2, options);
            webView2Environment = await CoreWebView2Environment.CreateAsync(null, tempWebCacheDir, options);

            //CoreWebView2 must be inialized before it can be used
            //wait for CoreWebView2 initialization
            //when complete, CoreWebView2Ready event will be thrown
            await wv.EnsureCoreWebView2Async(webView2Environment);

            //set source to 'about:blank'
            wv.Source = new Uri("about:blank", UriKind.Absolute);

            if (!String.IsNullOrEmpty(initialSource))
            {
                //Navigate to URL
                Debug.Print("Navigating to URL...");
                WebsiteNavigate(wv, initialSource);
            }
        }

        public void WebsiteNavigate(WebView2 wv, string dest)
        {
            string tempDest = dest;

            if (wv != null && wv.CoreWebView2 != null)
            {
                if (!String.IsNullOrEmpty(dest))
                {
                    if (dest != "about:blank" &&
                        !dest.StartsWith("edge://") &&
                        !dest.StartsWith("file://") &&
                        !dest.StartsWith("http://") &&
                        !dest.StartsWith("https://"))
                    {
                        //URL must start with one of the specified strings
                        //if not, pre-pend with "http://" or "https://"
                        //Debug.Print("Prepending \"http://\" to URL.");

                        //set value
                        tempDest = "http://" + dest;
                    }

                    //option 1
                    wv.CoreWebView2.Navigate(tempDest);

                    //option 2
                    //wv.Source = new Uri(tempDest, UriKind.Absolute);
                }
            }
        }
        private void WebView1_CoreWebView2Ready(object sender, EventArgs e)
        {
            Debug.Print("WebView1_CoreWebView2Ready");

            //subscribe to events (add event handlers)

            //ToDo: add any desired CoreWebView2 event handlers here
            //CoreWebView2 event handlers need to be added here
            //as this is the only guarantee that CoreWebView2 
            //initialization has completed.
            //WebView2 event handlers may be added here, as well

            //add CoreWebView2 event handlers
            WebView1.CoreWebView2.HistoryChanged += CoreWebView1_HistoryChanged;
            WebView1.CoreWebView2.NavigationCompleted += CoreWebView1_NavigationCompleted;
            WebView1.CoreWebView2.NavigationStarting += CoreWebView1_NavigationStarting;
            WebView1.CoreWebView2.WebMessageReceived += CoreWebView1_WebMessageReceived;

            //add WebView2 event handlers
            WebView1.NavigationCompleted += WebView1_NavigationCompleted;
            WebView1.NavigationStarting += WebView1_NavigationStarting;
            WebView1.WebMessageReceived += WebView1_WebMessageReceived;
        }


        private void CoreWebView1_HistoryChanged(object sender, object e)
        {
            Debug.Print("CoreWebView1_HistoryChanged");
        }

        private void CoreWebView1_NavigationCompleted(object sender, CoreWebView2NavigationCompletedEventArgs e)
        {
            Debug.Print("CoreWebView1_NavigationCompleted");
        }

        private void CoreWebView1_NavigationStarting(object sender, CoreWebView2NavigationStartingEventArgs e)
        {
            Debug.Print("CoreWebView1_NavigationStarting");
        }

        private void CoreWebView1_WebMessageReceived(object sender, CoreWebView2WebMessageReceivedEventArgs e)
        {
            Debug.Print("CoreWebView1_WebMessageReceived");
        }

        private void WebView1_NavigationCompleted(object sender, CoreWebView2NavigationCompletedEventArgs e)
        {
            Debug.Print("WebView1_NavigationCompleted");
        }

        private void WebView1_NavigationStarting(object sender, CoreWebView2NavigationStartingEventArgs e)
        {
            Debug.Print("WebView1_NavigationStarting");
        }

        private void WebView1_WebMessageReceived(object sender, CoreWebView2WebMessageReceivedEventArgs e)
        {
            Debug.Print("WebView1_WebMessageReceived");
        }


    }
}
```

This issue is also described in #340. 

[AB#28322080](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/28322080)

**Comment by cgeier** [comment]
One additional thing to note is that the WebView2 control property can't be set to null in the Visual Studio form designer, because it's a Uri. A Uri can't be set to null.

**Comment by mikemeinz** [comment]
I pulled the WebView2 initialization code out of the designer module and put it into the main program. I found that setting the Source property before initialization is totally complete, even when set to the default value, caused exceptions to occur randomly. My initialization code now doesn't set the Source property until the CoreWebView2Ready event is raised.

**Comment by champnic** [comment]
@cgeier This behavior (implicit initialization using .Source property) is By Design, but currently looks like a gap in our documentation: 

**Winforms WebView2.Source**: missing the information about implicit initialization
https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/winforms/0-9-515/microsoft-web-webview2-winforms-webview2#source

**WPF WebView2.Source**: has the information about implicit initialization
https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/wpf/0-9-515/microsoft-web-webview2-wpf-webview2#source

We'll work to get that documentation fixed. This also seems to be exacerbated by the fact that Winforms designer sets that property by default. We'll also see if we can get the source property to be set to null to allow it to be uninitialized when added in designer.

**Comment by champnic** [comment]
@mikemeinz Did you open issues tracking those failures?

**Comment by mikemeinz** [comment]
@champnic
I have Issue #344 opened. Others have reported similar experiences so I didn't open yet another issue.

**Comment by champnic** [comment]
Perfect, we're looking at that one. Thanks @mikemeinz!

**Comment by cgeier** [comment]
It seems leaving the Source property uninitialized, by removing

`this.WebView1.Source = new System.Uri("about:blank", System.UriKind.Absolute);`

 from the designer code (Form1.designer.cs), may be the way to go.

**Comment by cgeier** [comment]
If the Source property is removed from the Designer code (ex: Form1.Designer.cs), please ensure that the documentation states that the Source property needs to be set prior to using the WebView2 control. Additionally, specify whether or not one needs to set the Source property to "about:blank" before setting it to something else.

It seems that setting the WebView2 control Source property, seems to complete the initialization.

**Comment by champnic** [comment]
There are a couple of ways to initialize the control and using the Source property shouldn't be required. You can explicitly initialize the control by using EnsureCoreWebView2Async() and then calling Navigate/NavigateToString when the control is ready.

**Comment by champnic** [comment]
This is now fixed in the latest 1.0.707-prerelease SDK. Thanks!
